REST saga 2/3: Add REST endpoints to all RPCs

guggero opened the pull request:
This is part 2 of 3 of the REST improvement saga which will ultimately replace #4141:

- [x] Part 1/3 (#4250): Cleanup proto files, split `WalletUnlocker` into own file, generate Swagger docs for all RPCs, fix comment styles (replace `///` with `//` and `/**` with `/*`).
- [x] Part 2/3 (This PR): Add REST endpoints and `grpc-gateway` code to all subservers and update `grpc-gateway` library.
- [ ] Part 3/3 (#4141): Add a WebSocket proxy to the REST listener to support streaming responses.

Fixes #4329.

Roasbeef commented:
First part in, can be rebased now.

Roasbeef commented on lnrpc/chainrpc/chainnotifier.proto:
```diff
@@ -14,7 +16,12 @@ service ChainNotifier {
-    rpc RegisterConfirmationsNtfn (ConfRequest) returns (stream ConfEvent);
+    rpc RegisterConfirmationsNtfn (ConfRequest) returns (stream ConfEvent) {
+        option (google.api.http) = {
+            post: "/v2/chainnotifier/register_confirmations_ntfn"
```
Could simplify these paths to: `register/confirmation`, `register/spend`, `register/blocks`.

Roasbeef commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -15,40 +16,68 @@ service WalletKit {
-    rpc NextAddr (AddrRequest) returns (AddrResponse);
+    rpc NextAddr (AddrRequest) returns (AddrResponse) {
+        option (google.api.http) = {
+            get: "/v2/wallet/address/next"
```
Since this modifies state, make it a POST?

Roasbeef commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -60,7 +89,11 @@ service WalletKit {
-    rpc PendingSweeps (PendingSweepsRequest) returns (PendingSweepsResponse);
+    rpc PendingSweeps (PendingSweepsRequest) returns (PendingSweepsResponse) {
+        option (google.api.http) = {
+            get: "/v2/wallet/sweeps"
```
Latest version in master has a list _and_ a pending sweeps call.

Roasbeef commented on lnrpc/watchtowerrpc/watchtower.proto:
```diff
@@ -1,18 +1,24 @@
-    rpc GetInfo (GetInfoRequest) returns (GetInfoResponse);
+    rpc GetInfo (GetInfoRequest) returns (GetInfoResponse) {
+        option (google.api.http) = {
+            get: "/v2/watchtower"
```
Perhaps add a `/server` prefix here to disambiguate from the client calls below?

Roasbeef commented on rpcserver.go:
```diff
@@ -766,18 +766,26 @@ func (r *rpcServer) Start() error {
 		return err
 	}
+
+	for _, subServer := range r.subServers {
```
👍

Roasbeef commented:
Ready to leave draft now?

Roasbeef commented on go.sum:
```diff
@@ -105,14 +107,18 @@ github.com/gogo/protobuf v1.1.1 h1:72R+M5VuhED/KujmZVcIquuo8mBgX4oVda//DQb3PXo=
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/golang/protobuf v1.3.1 h1:YF8+flBXS5eO826T4nzqPrxfhQThhXl0YzfuUPu4SBg=
 github.com/golang/protobuf v1.3.1/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
+github.com/golang/protobuf v1.3.2 h1:6nsPYzhq5kReh6QImI3k5qWzO4PEbvbIW2cwSfR/6xs=
```
`go mod tidy` to get rid of these duplicates?

guggero replied:
Not all of them. Even if I remove the lines with the old versions, a `go get ./...` adds them back in. I assume it's a dependency that itself depends on an older version of `protobuf`.

Roasbeef commented on lnrpc/routerrpc/router.proto:
```diff
@@ -40,49 +55,77 @@ service Router {
-        returns (QueryProbabilityResponse);
+        returns (QueryProbabilityResponse) {
+        option (google.api.http) = {
+            get: "/v2/router/mc/probability/{from_node}/{to_node}/{amt_msat}"
```
👍

guggero commented:
Reworked the PR to implement the suggestion in #4329 and added a very simple check to `make rpc-check` that makes sure every new RPC is added to the `lnrpc/rest-annotations.yaml` file.

cfromknecht commented on Makefile:
```diff
@@ -210,6 +210,7 @@ rpc-format:
 
 rpc-check: rpc
 	@$(call print, "Verifying protos.")
+	for rpc in $$(find lnrpc/ -name "*.proto" | $(XARGS) awk '/    rpc /{print $$2}'); do if ! grep -q $$rpc lnrpc/rest-annotations.yaml; then echo "RPC $$rpc not added to lnrpc/rest-annotations.yaml"; exit 1; fi; done
```
👍
